Doc: fixed wrong description of PBKEYLEN conflict resolution (#2513)
Haivision/srt: docs/API/API-socket-options.md

  - **Bidirectional in Caller-Listener arrangement**: it is recommended to use
  a rule whereby you will be setting the `PBKEYLEN` exclusively either on the
  Listener (the service defines the encryption rules strictly) or on the Caller
  (the service allows all clients to freely decide about encryption, or the
  server uses more elaborate rules for encryption).

>   - **Bidirectional in Rendezvous mode**: In Rendezvous mode it is not possible for a party to be "unknown", 

Reviewer: > In Rendezvous mode it is not possible for a party to be "unknown"

What is meant here?

Author: When you establish a connection between two network endpoints, in case of caller-listener, it might be that the person who performs the connectio (or also an automated actor) is in control of only the caller side. In case of Rendezvous, the person who establishes the connection must be in control of both sides at once. This is because a "dedicated address" in case of caller-listener is only used by the caller (listener accepts connections from wherever they come), while in case of Rendezvous both sides use dedicated addresses (of the peer of one another).

This means that while a listener could have set PBKEYLEN to some value forcefurlly and might be unaware what the caller would do, the owner of the rendezvous connection knows this setting on both sides and knows what is happening on both sides - so none of the connection parties could be anyhow "unaware" of what the other party has set.

Author: Maybe better said would be something like: In Rendezvous mode it is assumed that someone who establishes a connection between two parties is in control of both of them,...

Reviewer 2: ```suggestion
  - **Bidirectional in Rendezvous mode**: In Rendezvous mode it is assumed that the settings (including `PBKEYLEN`) are known at both ends. 
```

  as it is with Listener mode, so both parties should set the same passphrase and the same key length, 
  or both should leave the `SRTO_PBKEYLEN` option unchanged (which results in default 16).

  - **Unwanted behavior cases**: if both parties set `PBKEYLEN` and the value
  on both sides is different, this is considered a conflict. It is resolved by
  the Initiator party, which takes its own value, if the `SRTO_SENDER`
  option is set to true. Otherwise, it takes the value from the Responder party. The

>   assignment of Initiator-Responder roles matches the Caller-Listener layout. In the case of Rendezvous the assignment is a matter of luck, resulting from the cookie comparison.

Author: But it's true. You can formulate it differently, but I think it should suffice to say "In the case of Rendezvous this assignment is according to the result of the cookie comparison and therefore unpredictable." @stevomatthews what do you think?

>   assignment of Initiator-Responder roles matches the Caller-Listener layout. In the case of Rendezvous the assignment is a matter of luck, resulting from the cookie comparison.

Reviewer: I don't like the word "luck" in the initial suggestion because it does not look scientific.
The cookie is even not a completely random thing, but rather an (MD5) function over time, host IP and port.

>   assignment of Initiator-Responder roles matches the Caller-Listener layout. In the case of Rendezvous the assignment is a matter of luck, resulting from the cookie comparison.

Author: That's perfectly right, but "random" isn't defined as "taken out of an unknown source", but "with no predictable result", and this here is the case. Hence I even support saying it directly that it's unpredictable.

>   assignment of Initiator-Responder roles matches the Caller-Listener layout. In the case of Rendezvous the assignment is a matter of luck, resulting from the cookie comparison.

Reviewer: The time is taken with a one-minute accuracy, making it quite predictable.

>   assignment of Initiator-Responder roles matches the Caller-Listener layout. In the case of Rendezvous the assignment is a matter of luck, resulting from the cookie comparison.

Author: Ok, I can make it extremely concise, like: "In the case of Rendezvous this assignment depends on the result of the cookie comparison". That's clear enough.

>   assignment of Initiator-Responder roles matches the Caller-Listener layout. In the case of Rendezvous the assignment is a matter of luck, resulting from the cookie comparison.

Reviewer 2: @ethouris  I think your last suggestion is fine.

  - **Bidirectional in Caller-Listener arrangement**: it is recommended to use
  a rule whereby you will be setting the `PBKEYLEN` exclusively either on the
  Listener (the service defines the encryption rules strictly) or on the Caller
  (the service allows all clients to freely decide about encryption, or the
  server uses more elaborate rules for encryption).

>   - **Bidirectional in Rendezvous arrangement**: In rendezvous there's no
>   situation of a potential "unknown party", as it is with listener side, so
>   both parties simply should set the same passphrase and the same key length,
>   or both should leave the `SRTO_PBKEYLEN` option unchanged (which results in
>   default 16).

Reviewer 2: ```suggestion
  - **Bidirectional in Rendezvous mode**: In Rendezvous mode it is not possible for a party to be "unknown", 
  as it is with Listener mode, so both parties should set the same passphrase and the same key length, 
  or both should leave the `SRTO_PBKEYLEN` option unchanged (which results in default 16).
```

Reviewer 2: @ethouris I prefer the term "mode" to "arrangement" when writing about Caller, Listener and Rendezvous. However, this is subjective. If we use "mode", then it should replace "arrangement" throughout the document.

Author: So far the practice was that "mode" was either of: Listener, Caller, Rendezvous - that is, touches upon only particular party. "Arrangement" was how both parties are organized together in order to form a connection - hence "Caller-Listener" or "Rendezvou" was an "arrangement".

  - **Unwanted behavior cases**: if both parties set `PBKEYLEN` and the value

>   on both sides is different, this is considered a conflict. It is resolved by
>   the Initiator party, which takes its own value, if it is set `SRTO_SENDER`
>   option to true, or takes the value from the Responder party otherwise. The
>   assignment of Initiator-Responder roles matches the Caller-Listener layout, and
>   in case of Rendezvous it's the matter of luck resulting from cookie comparison.
>   **It is highly recommended to never allow this to happen**, as this may also
>   override one party's setting of length 32 by the other one's 16.

Reviewer 2: ```suggestion
  on both sides is different, this is considered a conflict. It is resolved by
  the Initiator party, which takes its own value, if the `SRTO_SENDER`
  option is set to true. Otherwise, it takes the value from the Responder party. The
  assignment of Initiator-Responder roles matches the Caller-Listener layout. In the case of Rendezvous the assignment is a matter of luck, resulting from the cookie comparison.
  **It is highly recommended to never allow this to happen**, as this may result in having
  one party's setting of length = 32 be overridden by the other party's setting of length = 16.
```

the TSBPD mode (see [`SRTO_TSBPDMODE`](#SRTO_TSBPDMODE)) or encryption will not
work. Setting `SRTO_MINVERSION` to 1.3.0 is therefore recommended.

> This flag in versions above 1.3.0 also influences the conflict resolution for
> `SRTO_PBKEYLEN` in case when this flag is forcefully set on both connection
> parties simultaneously.

Reviewer 2: ```suggestion
This flag in versions above 1.3.0 also influences the conflict resolution for
`SRTO_PBKEYLEN` in the case where this flag is forcefully set on both connection
parties simultaneously.
```
